transcom/mymove
# MAIN B-21073

### **NOTE**
I apologize in advance for this PR - it is quite a mess. There are three separate INT commits and I had to make commits only in INT due to some differences in payload/model conversions and then those were merged into `main` and had to handle dealing with those changes in here.

The first INT PR also held changes for the Go upgrade, so... just do your best. If all tests are passing then all should be well, but it might be worth testing this locally just to be sure we are good to go.

This PR also contains a commit to fix a flaky test [from this PR](https://github.com/transcom/mymove/pull/13837) - with the addition of country checks this is always failing during CircleCI builds, so it is needed in order to pass `server-test`

[INT PR 1](https://github.com/transcom/mymove/pull/13853)
[INT PR 2](https://github.com/transcom/mymove/pull/13890)
[INT PR 3](https://github.com/transcom/mymove/pull/13902)

## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-21073)

## Summary

For the last twenty years I have been working on this BL item and its beautiful personality and quirks. After fighting a timeout issue likely due to so many concurrency issues happening when building addresses (and also adding countries to the db), I updated the `db-truncate` file to NOT clear the test db of `re_countries` data so we can retain that and avoid unnecessary inserts into the test database when building test data.

OKAY.

So the point of this BL item was to add a single source of truth for country information, which required a `re_countries` table with populated data. 

In addition to adding that data, we had to then refactor how we are handling country data in the backend and adding the `Country` object to the struct and then making that data accessible to the UI.

This PR does the following:
- adds migration file that adds the `re_countries` table and refactors the `addresses` table to use the `id` column of the `re_countries` in the new `country_id` column - drops the `country` column
- updates relevant Go structs
- adds/updates a buncha `Eager`/`EagerPreload` instances to use `"Addresses.Country"` now that the `Country` object holds the necessary data
- updates the `Address` definition for Swagger to only be two characters (we will handle this later on when international rolls out)
- defaults all created addresses to have a `US` country
- updates all tests (roughly 3 million tests)

### How to test

1. Run `make server_run` and see the migrations run
2. Check your database and you should now see a `re_countries` table, as well as the `addresses` table having a removed `country` column and added `country_id` column
3. Open up the application and console log some addresses
4. You should be able to see the `Country` object attached to the addresses
5. Check any situation/process where `Country` is important


## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-21073)

## Summary

This is adding a little cleanup + one missed oopsie of validating the country when the Prime attempts to create an address that contains a country code that is not supported.

Since we do not support international moves at this time, I added in a check that IF the Prime provides a country code that is not `US`, then we are going to send back an error message saying the country code they provided is not supported.

### How to test

1. Access MM as Prime Sim
2. Find a move that you can create a shipment for
3. Endpoint would be `/prime/v3/mto-shipments`
4. Try to pass in an address that has a `Country` value of something that is NOT `US` - should get back a validation error
```
{
  "moveTaskOrderID": "0f21be2a-5045-4e90-8212-95a92de90056",
  "requestedPickupDate": "2024-10-12",
  "primeEstimatedWeight": 4000,
  "customerRemarks": "handle with care",
  "pickupAddress": {
    "streetAddress1": "123 OtherWorldly Ave",
    "city": "Not a US Country",
    "state": "AK",
    "postalCode": "90210",
    "country": "GB"
  },
  "destinationAddress": {
    "streetAddress1": "123 Main Ave",
    "streetAddress2": "Apartment 9000",
    "streetAddress3": "Montmârtre",
    "city": "Anytown",
    "state": "AL",
    "postalCode": "90210",
    "country": "US"
  },
  "shipmentType": "HHG"
}
```

![Screenshot 2024-10-11 at 3 55 00 PM](https://github.com/user-attachments/assets/4407105d-b968-4584-8291-1dba571e536d)



## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-21073)

## Summary

Discovered that the service member profile creation/update flow does not use the expected service objects and instead uses a function in the models file. Updating that function to include `country_id` when saving the service member.

### How to test

1. Access MM as a customer
2. Create your profile & your addresses
3. Check out the `addresses` table and ensure that the `country_id` is populated
